File: 02-dynamic-programming/10_longest_palindrome.c
#include "../test_runner.h"

/* Write result into out buffer */
void longest_palindrome(char* s, char* out) {
    out[0] = '\0';
}

int main() {
    char out[256];
    longest_palindrome("babad", out);
    int len1 = strlen(out);
    TEST_BOOL("babad len>=3", len1 >= 3, true);

    longest_palindrome("cbbd", out);
    int len2 = strlen(out);
    TEST_BOOL("cbbd len>=2", len2 >= 2, true);

    longest_palindrome("a", out);
    TEST_STR("single char", out, "a");

    longest_palindrome("racecar", out);
    TEST_STR("full palindrome", out, "racecar");

    TEST_SUMMARY();
}

File: 02-dynamic-programming/10_longest_palindrome.py
import sys, os; sys.path.append(os.path.join(os.path.dirname(os.path.abspath(__file__)), '..'))
from test_runner import run_tests


# Longest Palindromic Substring (LC #5) -- Medium
# Find longest palindromic substring.
#
#   Approach 1: Expand around center for each index. O(n^2) time, O(1) space.
#   Approach 2: DP table `dp[i][j]` = is s[i..j] palindrome?

def longest_palindrome(s):
    """Return the longest palindromic substring"""
    pass


# Multiple valid answers possible, so check length and palindrome property
def validate(s):
    result = longest_palindrome(s)
    if result is None:
        return False
    return result == result[::-1]


run_tests(validate, [
    (("babad",), True),
    (("cbbd",),  True),
    (("a",),     True),
    (("racecar",), True),
])
